src/protocol.rs

```diff
@@ -300,24 +281,50 @@ impl ObliviousDoHMessagePlaintext for ObliviousDoHResponseBody {
         &self.padding
     }
 }
+#[derive(SerdeSerialize, SerdeDeserialize, Clone, Debug)]
+pub struct ResponseNonce(Vec<u8>);
+impl_custom_serde!(ResponseNonce);
 
-/// Derives a key and nonce pair using the odoh secret.
-fn derive_secrets(odoh_secret: &[u8], query: &ObliviousDoHQueryBody) -> Result<(Vec<u8>, Vec<u8>)> {
+/// Derives a key and nonce pair using the odoh secret and response_nonce.
+/// If response_nonce is already specified, uses it to generate the key and nonce
+/// Otherwise, response_nonce is generated, and response_nonce is returned with key and nonce
+fn derive_secrets(
+    odoh_secret: &[u8],
+    query: &ObliviousDoHQueryBody,
+    response_nonce: Option<Vec<u8>>,
+) -> Result<(Vec<u8>, Vec<u8>, ResponseNonce)> {
     let key_info = LABEL_KEY.to_vec();
     let nonce_info = LABEL_NONCE.to_vec();
     let query_bytes = query.to_bytes()?;
+    let response_nonce = if let Some(r) = response_nonce {
+        if r.len() != RESPONSE_NONCE_SIZE {
+            return Err(anyhow!(
+                "response_nonce is smaller than max(key, nonce) size"
+            ));
+        }
+        ResponseNonce(r)
+    } else {
+        // this only works for RESPONSE_NONCE_SIZE <= 32
+        // see: https://docs.rs/rand/0.8.3/rand/distributions/struct.Standard.html
```

Comment:
Does this generate a cryptographically secure random value? The text says:

> A generic random value distribution, implemented for many primitive types. Usually generates values with a numerically uniform distribution, and with a range appropriate to the type.

... which suggests to me it might not be safe for the purposes of a nonce? I don't know what is best practice here, but maybe we can lean on [ring](https://briansmith.org/rustdoc/ring/rand/index.html)?

Reply:
This uses `thread_rng` which is cryptographically secure: https://docs.rs/rand/0.8.3/rand/rngs/struct.ThreadRng.html

Reply:
Perfect 👍 Thanks!